frontend/dockerfile/instructions/parse.go

```diff
@@ -341,8 +343,14 @@ func parseWorkdir(req parseRequest) (*WorkdirCommand, error) {
 	if err != nil {
 		return nil, err
 	}
+
+	path, err := system.CheckSystemDriveAndRemoveDriveLetter(req.args[0])
+	if err != nil {
+		return nil, err
+	}
+
 	return &WorkdirCommand{
-		Path:            req.args[0],
+		Path:            filepath.ToSlash(path),
```

Comment:
Maybe instead of using `filepath` and making decisions based on the native architecture, it would be better to move this to dockerfile2llb and use the platform of the dispatchstate https://github.com/moby/buildkit/blob/3f3957d16c5818801986f5b0db4599f5c419f1ce/frontend/dockerfile/dockerfile2llb/convert.go#L546 to make the decision about converting paths. Would like that windows dockerfile that doesn't use `RUN` would work on Linux as well, like it works today if unix paths are used.

This looks somewhat similar to https://github.com/moby/buildkit/issues/1560 that you also might want to take.

Reply:
To be clear, is the idea that `WORKDIR C:\test` should work when running a Linux-based build that is targeting the Windows platform? That seems reasonable, but will require a little bit more learning on my part. I like this more, because one of the outstanding concerns I had with my tests was the runtime platform check. I can split the current test into two, and confirm that we get rational behaviour for both Linux and Windows targets, irrespective of the execution platform.

Reply:
> To be clear, is the idea that WORKDIR C:\test should work when running a Linux-based build against the Windows targetPlatform? 

Yes. This works today (outside Moby integration). But I think today you just always need to use unix paths that should work even if you build for windows. Normally, I would just always recommend using unix paths in Dockerfiles but because old builder already supports windows paths BuildKit should as well. But the same Dockerfile should be usable for building natively on Windows and cross-building on Linux.

> I assume WORKDIR C:\test should fail if the targetPlatform is Linux...

Yes, we shouldn't try to do any normalization magic for linux targets. If target is windows we can check if it is a windows style path and normalize to unix path if needed.